### PR TITLE
Refactor CSRF checks

### LINKED ISSUE #71: Use Origin header if Referrer is not present for CSRF checks
See: https://support.joomlatools.com/forums/topic/6964
See: https://support.joomlatools.com/forums/topic/7009
See: https://support.joomlatools.com/forums/topic/7038
See: https://support.joomlatools.com/forums/topic/7230
See: https://support.joomlatools.com/forums/topic/6717

If PHP is configured to run byhind a proxy the HTTP_REFERER header might not be available which leads to during POST requests. To reoslve this add HTTP_ORIGIN header to mitigate CSRF attacks.

More info see: https://www.owasp.org/index.php/CSRF_Prevention_Cheat_Sheet#Standard_Header_Checks

- [x] Add KDispatcherRequestInterface::getOrigin
- [x] Check for internal URL in both KDispatcherRequestInterface::getOrigin and KDispatcherRequestInterface::getReferrer
- [x] Remove KFilterInternalUrl since the check is moved into the request object
- [x] Check for referrer and fallback to origin header in CSRF authenticator

### PR DESCRIPTION
**Pull request by ercanozkaya**
* Add KDispatcherRequestInterface::getOrigin
* Check for internal URL in both KDispatcherRequestInterface::getOrigin and KDispatcherRequestInterface::getReferrer
* Remove KFilterInternalUrl since the check is moved into the request object
* Check for referrer and fallback to origin header in CSRF authenticator

Closes #71

### CODE REVIEW

**Comment by ercanozkaya** [comment]
@johanjanssens can you review please?

File: code/libraries/joomlatools/library/dispatcher/request/abstract.php [hunk 1]
```diff
@@ -600,16 +600,59 @@ public function getReferrer($isInternal = true)
...
+     * Returns the HTTP origin header.
+     *
+     * @param   boolean  $isInternal Only allow internal URLs
+     * @return  KHttpUrl|null  A HttpUrl object or NULL if no origin header could be found
+     */
+    public function getOrigin($isInternal = true)
+    {
+        $origin = null;
+
+        if ($this->_headers->has('Origin'))
```
Comment (johanjanssens) [inline code/libraries/joomlatools/library/dispatcher/request/abstract.php]: @ercanozkaya Should we filter Origin and consider it unsafe? See: https://github.com/timble/kodekit/issues/137

Comment (ercanozkaya) [inline code/libraries/joomlatools/library/dispatcher/request/abstract.php]: @johanjanssens You are right we should, I added a call to the filter URL. Can you please check again?

**Comment by johanjanssens** [comment]
@ercanozkaya We have a setReferrer in request to be able to set it after it has been cleaned. This prevents the need to re-clean it. 

Code does something like: if(!isset($this->_referrer) && ($this->_headers->has('Referer') || $this->data->has('_referrer'))) {

Should we try to do the same for Origin to maintain consistency? Add a setOrigin and additional check so we only clean it once? The setOrigin shouldn't go into the interface.

**Comment by johanjanssens** [comment]
@ercanozkaya Can you merge this with Kodekit and created a related ticket?

**Pull request merged**